lib/propolis/src/block/file.rs

```diff
@@ -238,30 +268,200 @@ impl block::Backend for FileBackend {
     }
 }
 
-/// Attempt to query the Write-Cache-Enable state for a given open device
-///
-/// Block devices (including "real" disks and zvols) will regard all writes as
-/// synchronous when performed via the /dev/rdsk endpoint when WCE is not
-/// enabled.  With WCE enabled, writes can be cached on the device, to be
-/// flushed later via fsync().
-fn get_wce(fp: &File) -> Option<bool> {
-    let ft = fp.metadata().ok()?.file_type();
-    if !ft.is_char_device() {
-        return None;
+mod dkioc {
+    #![allow(non_camel_case_types)]
+
+    use std::fs::File;
+    use std::io::Result;
+    use std::os::raw::{c_int, c_longlong, c_uint};
+    use std::os::unix::fs::FileTypeExt;
+    use std::os::unix::io::AsRawFd;
+
+    use crate::block::DEFAULT_BLOCK_SIZE;
+    use crate::util::ioctl;
+
+    const DKIOC: i32 = 0x04 << 8;
+    const DKIOCGETWCE: i32 = DKIOC | 36;
+    const DKIOCSETWCE: i32 = DKIOC | 37;
+    const DKIOCGMEDIAINFOEXT: i32 = DKIOC | 48;
+    const DKIOCFREE: i32 = DKIOC | 50;
+    const DKIOC_CANFREE: i32 = DKIOC | 60;
+
+    #[derive(Copy, Clone)]
+    pub(crate) enum DiscardMech {
+        /// Discard via `ioctl(DKIOCFREE)`
+        DkiocFree,
+        /// Discard via `fcntl(F_FREESP)`
+        FnctlFreesp,
     }
 
-    let mut res: c_int = 0;
-    let _ = unsafe {
-        ioctl(fp.as_raw_fd(), DKIOCGETWCE, &mut res as *mut c_int as _).ok()?
-    };
-    Some(res != 0)
-}
+    #[derive(Copy, Clone)]
+    #[allow(unused, dead_code)]
+    pub(crate) struct DiskInfo {
+        /// WCE state (if any) for disk
+        ///
+        /// Block devices (including "real" disks and zvols) will regard all writes as
+        /// synchronous when performed via the /dev/rdsk endpoint when WCE is not
+        /// enabled.  With WCE enabled, writes can be cached on the device, to be
+        /// flushed later via fsync().
+        pub wce_state: Option<bool>,
+        /// Block size of disk
+        pub block_size: u32,
+        /// Does the disk support use of DKIOCFREE or F_FREESP?
+        pub discard_mech: Option<DiscardMech>,
+    }
+    impl Default for DiskInfo {
+        fn default() -> Self {
+            Self {
+                wce_state: None,
+                block_size: DEFAULT_BLOCK_SIZE,
+                discard_mech: None,
+            }
+        }
+    }
+
+    pub(crate) fn disk_info(fp: &File) -> DiskInfo {
+        match fp.metadata() {
+            Ok(ft) if ft.file_type().is_char_device() => {
+                // Continue on to attempt DKIOC lookups on raw disk devices
+            }
+            Ok(ft) if ft.file_type().is_file() => {
+                // Assume fcntl(F_FREESP) support for files
```

Comment:
Optional, up to you: I would be somewhat tempted to put the `target_os = "illumos"` check here (returning `None` if it doesn't hold) and then marking the F_FREESP case as `unreachable!` on other OSes below instead of using the invalid `fcntl` command code.